bee-network/bee-autopeering/src/request.rs

```diff
@@ -0,0 +1,202 @@
+        &mut self,
+        peer_id: PeerId,
+        peer_addr: IpAddr,
+        response_tx: Option<ResponseTx>,
+    ) -> VerificationRequest {
+        let key = RequestKey {
+            peer_id,
+            request_id: TypeId::of::<VerificationRequest>(),
+        };
+
+        let verif_req = VerificationRequest::new(self.version, self.network_id, self.source_addr, peer_addr);
+        let timestamp = verif_req.timestamp();
+
+        let request_hash = msg_hash(
+            MessageType::VerificationRequest,
+            &verif_req.to_protobuf().expect("error encoding verification request"),
+        );
+
+        let value = RequestValue {
+            request_hash,
+            expiration_time: timestamp + REQUEST_EXPIRATION_SECS,
+            response_tx,
+        };
+
+        let _ = self.open_requests.insert(key, value);
+
+        verif_req
+    }
+
+    pub(crate) fn new_discovery_request(
+        &mut self,
+        peer_id: PeerId,
+        response_tx: Option<ResponseTx>,
+    ) -> DiscoveryRequest {
+        let key = RequestKey {
+            peer_id,
+            request_id: TypeId::of::<DiscoveryRequest>(),
+        };
+
+        let disc_req = DiscoveryRequest::new();
+        let timestamp = disc_req.timestamp();
+
+        let request_hash = msg_hash(
+            MessageType::DiscoveryRequest,
+            &disc_req.to_protobuf().expect("error encoding discovery request"),
+        );
+
+        let value = RequestValue {
+            request_hash,
+            expiration_time: timestamp + REQUEST_EXPIRATION_SECS,
+            response_tx,
+        };
+
+        let _ = self.open_requests.insert(key, value);
+
+        disc_req
+    }
+
+    pub(crate) fn new_peering_request(
+        &mut self,
+        peer_id: PeerId,
+        response_tx: Option<ResponseTx>,
+        local: &Local,
+    ) -> PeeringRequest {
+        let key = RequestKey {
+            peer_id,
+            request_id: TypeId::of::<PeeringRequest>(),
+        };
+
+        let peer_req = PeeringRequest::new(local.public_salt().expect("missing public salt"));
+
+        let timestamp = peer_req.timestamp();
+
+        let request_hash = msg_hash(
+            MessageType::PeeringRequest,
+            &peer_req.to_protobuf().expect("error encoding peering request"),
+        );
+
+        let value = RequestValue {
+            request_hash,
+            expiration_time: timestamp + REQUEST_EXPIRATION_SECS,
+            response_tx,
+        };
+
+        let _ = self.open_requests.insert(key, value);
+
+        peer_req
+    }
+
+    pub(crate) fn pull<R: Request + 'static>(&mut self, peer_id: &PeerId) -> Option<RequestValue> {
+        let key = RequestKey {
+            peer_id: *peer_id,
+            request_id: TypeId::of::<R>(),
+        };
+
+        self.open_requests.remove(&key)
+    }
+}
+
+pub(crate) fn is_expired(timestamp: Timestamp) -> bool {
+    time::since(timestamp).map_or(false, |ts| ts >= REQUEST_EXPIRATION_SECS)
```

Comment:
Just making sure false is a sensible default here

Reply:
Good point. I added a `Note` to explain it.

Assume the function is called with a `timestamp` representing some point in the future. Then `time::since` returns `None` for obvious reasons, and has to be mapped to `false` (not expired).